File: examples/js-examples/Collision.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

using meshA = ml.makeUVSphere(1.0, 16, 16); // make mesh A
using meshB = ml.makeUVSphere(1.0, 16, 16); // make mesh B
// shift mesh B for better demonstration
using shift = ml.AffineXf3f.translation({ x: 0.1, y: 0.1, z: 0.1 });
meshB.transform(shift);

// find each pair of colliding faces
const collidingFacePairs = ml.findCollidingTriangles(meshA, meshB);
for (const { aFace, bFace } of collidingFacePairs)
  console.log(`${aFace} ${bFace}`);

// find bitsets of colliding faces
const bitsets = ml.findCollidingTriangleBitsets(meshA, meshB);
using collidingFacesA = bitsets.a;
using collidingFacesB = bitsets.b;
console.log(collidingFacesA.count()); // number of colliding faces from mesh A
console.log(collidingFacesB.count()); // number of colliding faces from mesh B

// fast check whether mesh A and mesh B collide
const isColliding = ml.findCollidingTriangles(meshA, meshB, true).length > 0;
console.log(isColliding);

File: examples/cpp-examples/Collision.dox.cpp
#include "MRMesh/MRMakeSphereMesh.h"
#include "MRMesh/MRMesh.h"
#include "MRMesh/MRMatrix3.h"
#include "MRMesh/MRAffineXf.h"
#include "MRMesh/MRMeshCollide.h"
#include <iostream>

int main()
{
    auto meshA = MR::makeUVSphere(); // make mesh A
    auto meshB = MR::makeUVSphere(); // make mesh B
    meshB.transform( MR::AffineXf3f::translation( MR::Vector3f( 0.1f, 0.1f, 0.1f ) ) ); // shift mesh B for better demonstration

    auto collidingFacePairs = MR::findCollidingTriangles( meshA, meshB ); // find each pair of colliding faces
    for ( const auto [fa, fb] : collidingFacePairs )
        std::cout << int( fa ) << " " << int( fb ) << "\n"; // print each pair of colliding faces

    auto [collidingFaceBitSetA, collidingFaceBitSetB] = MR::findCollidingTriangleBitsets( meshA, meshB ); // find bitsets of colliding faces
    std::cout << collidingFaceBitSetA.count() << "\n"; // print number of colliding faces from mesh A
    std::cout << collidingFaceBitSetB.count() << "\n"; // print number of colliding faces from mesh B


    auto isColliding = !MR::findCollidingTriangles( meshA, meshB, nullptr, true ).empty(); // fast check if mesh A and mesh B collide
    std::cout << isColliding << "\n";
    return 0;
}
